File: src/public/dependency.rs
use std::path::PathBuf;

use serde::Deserialize;
use serde::Serialize;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Dependency {
  /// Identifier of the import
  pub specifier: String,
  pub specifier_type: SpecifierType,
  /// Whether the dependency is an entry
  pub is_entry: bool,
  /// When the dependency should be loaded
  pub priority: DependencyPriority,
  /// Path to the file that imported this dependency
  pub source_path: PathBuf,
  /// Path to resolve the specifier from
  pub resolve_from: PathBuf,
  /// Symbols that are imported from this path
  pub imported_symbols: Vec<ImportSymbol>,
  /// Where to place the dependency within the bundle
  pub bundle_behavior: BundleBehavior,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SpecifierType {
  ESM,
  Commonjs,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DependencyPriority {
  /// Static import
  Sync,
  /// Dynamic import
  Lazy,
}


#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash)]
pub enum ImportSymbol {
  // import './foo'
  Unnamed,
  // import { foo } from './foo'
  // import { foo: bar } from './foo'
  Named(String),
  // import foo from './foo'
  Default,
  // import * as foo from './foo'
  Namespace(String),
  // export * from './foo'
  Reexport,
  // import('./foo')
  Dynamic,
  // require('./foo')
  Commonjs,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum BundleBehavior {
  Inline,
  Isolated,
}

File: deno/adapter/javascript/dependency.js
export class Dependency {
  #ref

  constructor(ref) {
    this.#ref = ref
  }

  get id() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 0)
  }
  
  get specifier() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 1)
  }
  
  get specifierType() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 2)
  }
  
  get isEntry() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 3)
  }
  
  get priority() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 4)
  }
  
  get sourcePath() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 5)
  }
  
  get sourceAsset() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 6)
  }
  
  get resolveFrom() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 7)
  }
  
  get importedSymbols() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 8)
  }
  
  get bundleBehavior() {
    return globalThis.Mach.ops.getter_dependency(this.#ref, 9)
  }
}
